[jit] Fix bugs in assignment to optionals

zdevito opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#24989 [jit] Fix bugs in assignment to optionals**
* #23281 Load tensors directly from pickle archive

This fixes the cases where a type annotated with optional cannot
be conditionally assigned to none:

```
x : Optional[int] = 4
if ...:
 x = None
```

Differential Revision: [D16949314](https://our.internmc.facebook.com/intern/diff/D16949314)

zdevito committed:
[jit] Fix bugs in assignment to optionals

This fixes the cases where a type annotated with optional cannot
be conditionally assigned to none:

```
x : Optional[int] = 4
if ...:
 x = None
```

zdevito commented on torch/csrc/jit/script/compiler.cpp:
```diff
@@ -947,13 +984,6 @@ struct to_ir {
         case TK_AUG_ASSIGN:
           emitAugAssignment(AugAssign(stmt));
           break;
-        case TK_GLOBAL:
```
this code was totally bogus stuff left over from caffe2 script!

jamesr66a replied:
![image](https://user-images.githubusercontent.com/4685384/63557116-331f4200-c4fc-11e9-996a-81011b095d95.png)

zdevito committed:
Update on "[jit] Fix bugs in assignment to optionals"

[jit] Fix bugs in assignment to optionals

This fixes the cases where a type annotated with optional cannot
be conditionally assigned to none:

```
x : Optional[int] = 4
if ...:
 x = None
```

gh-metadata: pytorch pytorch 24989 gh/zdevito/90/head

zdevito committed:
Update on "[jit] Fix bugs in assignment to optionals"

[jit] Fix bugs in assignment to optionals

This fixes the cases where a type annotated with optional cannot
be conditionally assigned to none:

```
x : Optional[int] = 4
if ...:
 x = None
```

gh-metadata: pytorch pytorch 24989 gh/zdevito/90/head

ZolotukhinM commented on torch/csrc/jit/script/compiler.cpp:
```diff
@@ -229,10 +229,14 @@ struct Environment {
   }
 
-  void insertStore(const std::string& name, const SourceRange& loc, Value* v) {
+  void insertStore(
```
Maybe add a comment demonstrating a case when we need to use type different than `store->input()->type()`.

ZolotukhinM commented on torch/csrc/jit/script/convert_to_ssa.cpp:
```diff
@@ -118,20 +118,25 @@ struct ControlFlowLoadStores {
-      // and we do not refine the type of the parent variable since the loop may
-      // not be entered.
+      // if the variable local to the loop body, then
+      // we do not need a loop carried variabl for it
```
typo: variabl

ZolotukhinM reviewed (approved):
Looks good to me! cc @eellison

eellison commented on torch/csrc/jit/script/convert_to_ssa.cpp:
```diff
@@ -118,20 +118,25 @@ struct ControlFlowLoadStores {
+      // of the loop and the input loop carried dependencies are conservatively
+      // the union of the output of the body and the input to the loop
+      auto block_type = loop_vars->findInThisFrame(name);
+      auto unified_type = unifyTypes(parent_type, block_type).value();
```
Why do we have to unify `parent_type` and `block_type` here ? In compiler.cpp there isn't a similar logic; we enforce that the loop block variables subtype the parent type and when the loop is done emitting the variable remains the parent type.

I think  having a difference in type between the type environment in compiler.cpp and the IR could be a source of bugs.

eellison replied:
I think this will compile and then fail in schema matching as a result of ^^^

```
def foo():
   a = 1
   for i in range(0):
       a = None
   print(a + a)
```

zdevito replied:
Because store doesn't remember the type of parent variable.
```
a : Optional[Tensor] = torch.ones()
if True:
  a = None
```

The store will store a value of type tensor. It doesn't remember that it is storing optional tensor (afterall, where would it?).  So the convert_to_ssa pass is going to derive type Tensor, but the original script has type Optional[Tensor].

eellison replied:
Makes sense. another option is making prim::Store remember Optional[Tensor] here

zdevito committed:
Update on "[jit] Fix bugs in assignment to optionals"

[jit] Fix bugs in assignment to optionals

This fixes the cases where a type annotated with optional cannot
be conditionally assigned to none:

```
x : Optional[int] = 4
if ...:
 x = None
```

gh-metadata: pytorch pytorch 24989 gh/zdevito/90/head

eellison commented on torch/csrc/jit/script/compiler.cpp:
```diff
@@ -306,8 +323,15 @@ struct Environment {
+          parent_type,
+          as_simple_value,
+          /*allow_conversions=*/true);
+      if (!as_simple_value->type()->isSubtypeOf(parent_type)) {
```
I think the only case where `tryConvertToType` will return a new type and that type will be a subtype of the parent type is with a None and an optional. Another option is to break out that part of the `tryConvertToType` function so this code will be more transparent.

zdevito replied:
We also want variable assignment to follow all the same type conversion rules as schemas, so using tryConvertToType is correct here.
